core/nonce.go

```diff
@@ -18,6 +19,7 @@ const MaxUsed = 65536
 
 // NonceService generates, cancels, and tracks Nonces.
 type NonceService struct {
+	mu       sync.Mutex
```

Comment:
Since `Valid` is a read-only operation it might make sense to use a `RWMutex` so we can do concurrent reads of the map.